### PR TITLE
add VpcRef ADT to allow VPC ID to be set more flexibly

### PR DESCRIPTION
**Pull request by bpholt**
This PR partially addresses issue #102 / #107  with regard to VPC ID references.

### CODE REVIEW

File: src/test/scala/com/monsanto/arch/cloudformation/model/FnIf_UT.scala [hunk 1]
```diff
@@ -18,7 +18,7 @@ class FnIf_UT extends FunSpec with Matchers {
         function = `Fn::Equals`(a = StringToken("true"), b = StringToken("false"))
       )
 
-      val vpcToken = UNSAFEToken[ResourceRef[`AWS::EC2::VPC`]]("vpc-b5f389d0")
```
Comment (seanmcl) [inline src/test/scala/com/monsanto/arch/cloudformation/model/FnIf_UT.scala]: Nice.

File: src/main/scala/com/monsanto/arch/cloudformation/model/resource/EC2.scala [hunk 2]
```diff
@@ -562,6 +562,35 @@ object `AWS::EC2::VPC` extends DefaultJsonProtocol {
   implicit val format: JsonFormat[`AWS::EC2::VPC`] = jsonFormat7(`AWS::EC2::VPC`.apply)
 }
 
+sealed trait VpcRef
+case class ResourceRefVpc(ref: Token[ResourceRef[`AWS::EC2::VPC`]]) extends VpcRef
+case class ImportedVpc(fn: `Fn::ImportValue`) extends VpcRef
+case class MappedVpc(fn: `Fn::FindInMap`[String]) extends VpcRef
+case class VpcLiteral(vpcId: String) extends VpcRef
+
+case class VpcId private(resource: VpcRef)
```
Comment (seanmcl) [inline src/main/scala/com/monsanto/arch/cloudformation/model/resource/EC2.scala]: What are you trying to prevent with the `private` here?

Comment (bpholt) [reply src/main/scala/com/monsanto/arch/cloudformation/model/resource/EC2.scala]: The idea is that one should never have to construct a `VpcId` explicitly. I was thinking users would create instances of `VpcRef` and they would be implicitly converted to `VpcId` instances by the implicit methods below.

I modeled this off some other code in this project (although I don't remember exactly where). Idk, maybe it's cleaner to get rid of `VpcId` and use `VpcRef` directly instead? That seems to work too.

Comment (seanmcl) [reply src/main/scala/com/monsanto/arch/cloudformation/model/resource/EC2.scala]: Seems fine to me.  Just curious if there were some illegal cases you were preventing.

**Comment by bkrodgers** [comment]
Nice solution to a problem I've wanted to tackle for awhile.  I bet there are quite a few other things that could benefit from this approach.  For now we'll merge as is, but might want to see if we can genericize it further to make it reusable for other types.

**Pull request merged**